plugins/modules/ec2_metadata_facts.py

```diff
@@ -484,11 +485,35 @@ def __init__(
         self._token = None
         self._prefix = "ansible_ec2_%s"
 
+    def decode_user_data(self, data):
+        is_compressed = False
+
+        # Check if data is compressed using zlib header
+        if data.startswith(b"\x78\x9c") or data.startswith(b"\x1f\x8b"):
+            is_compressed = True
+
+        try:
+            if is_compressed:
+                # Data is compressed, attempt decompression and decode using UTF-8
+                decoded_data = zlib.decompress(data, zlib.MAX_WBITS | 32).decode("utf-8")
```

Comment:
If the decompression succeeds, but the decoding fails with a `UnicodeDecodeError`, then the `except UnicodeDecodeError` block is going to fail because it will try to decode compressed data. The easiest way to handle this is probably with two try/except blocks, one to decompress the data, and the other to decode it.